pkg/dab/find.go

```diff
@@ -0,0 +1,237 @@
+// For example, 'ActionableSearch_Any' may cause the search to return a formula or a module;
+// whereas if you only want only want to detect modules, you should use 'ActionableSearch_Module' instead.
+// See the comments on the constants for details on other options.
+//
+// It is possible for a module and plot to be returned;
+// or a module alone; or a plot alone; or a formula alone;
+// or an error; or all four of them may be nil at once if the search found nothing.
+// The foundPath and remainingSearchPath values are always returned,
+// even in the case of errors.
+//
+// Errors:
+//
+//  - warpforge-error-datatoonew -- when found data is not supported by this version of warpforge
+//  - warpforge-error-invalid-argument -- when a file contains the wrong content for its filename
+//  - warpforge-error-io -- when an IO error occurs during search
+//  - warpforge-error-missing -- when an expected file is missing
+//  - warpforge-error-module-invalid -- when a read module contains invalid data
+//  - warpforge-error-searching-filesystem -- when the search of the filesystem produces an invalid result
+//  - warpforge-error-serialization -- when IPLD deserialization fails
+func FindActionableFromFS(
+	fsys fs.FS,
+	basisPath string, searchPath string, searchUp bool,
+	accept ActionableSearch,
+) (
+	m *wfapi.Module, p *wfapi.Plot, f *wfapi.Formula,
+	foundPath string, remainingSearchPath string, err error,
+) {
+	remainingSearchPath = searchPath
+
+	// First round: this may be a file, and may be one of any of the types.
+	// Stat it and consider that first:
+	// if it is a file, that's its own whole detection procedure (and means no further search);
+	// if it's not a file, we can procede to the behavior for dir feature detection (without popping).
+	foundPath = filepath.Join(basisPath, remainingSearchPath)
+	fi, e2 := stat(fsys, foundPath)
+	if e2 != nil {
+		err = e2
+		return
+	}
+	if fi.Mode()&^fs.ModePerm == 0 {
+		fh, e2 := open(fsys, foundPath)
+		defer fh.Close()
+		if e2 != nil {
+			err = e2
+			return
+		}
+		m, p, f, err = findActionableFromFile(fh, accept)
+		if m != nil { // If a module was found, a plot may also exist as a sibling file.
+			p, e2 = PlotFromFile(fsys, filepath.Join(filepath.Dir(filepath.Join(basisPath, remainingSearchPath)), "plot.wf"))
+			if !errors.Is(e2, fs.ErrNotExist) {
+				err = wfapi.ErrorSearchingFilesystem("loading plot associated with a module", e2)
+			}
+		}
+		return
+	}
+
+	// Iteratively check directories.
+	// In each directory, look for well-known file names.
+	// If we find a module, that dominates; then plot; then formula.
+	// When there's no match, if searchUp is enabled, then pop one segment off searchPath and search again.
+	for {
+		// Peek for module.
+		if accept&ActionableSearch_Module > 0 {
+			foundPath = filepath.Join(basisPath, remainingSearchPath, "module.wf")
+			m, e2 = ModuleFromFile(fsys, foundPath)
+			if serum.Code(e2) != wfapi.ECodeMissing { // notexist is ignored.
```

Comment:
This probably resolves to `true` for a `nil` error (`Code` would return an empty string). I don't think this describes the expected behavior. Probably need tests for this.